### ISSUE TITLE
Async named functions break

### ISSUE THREAD

**Issue by benjamingr** [babel/babel#979]
I ran the following code:
```js
    await Promise.map(images, async function rt(image,_,_2,ret){
    	try{
    		await ftp.putAsync(image,"/site/wwwroot/www/" + image);
    	} catch(e){
    		if(ret) throw new Exception("can't upload");
    		return await rt(image,_, _2, true);
    	}
    	console.log("uploaded", image)
    });
```

Which gives me `ReferenceError: rt is not defined`

**Comment by sebmck** [comment]
[Works for me](http://babeljs.io/repl/#?experimental=true&playground=false&evaluate=true&loose=false&spec=false&code=var%20foo%20%3D%20async%20function%20bar()%20%7B%0A%20%20console.log(bar%20%2B%20%22%22)%3B%0A%7D%3B%0A%0Afoo()%3B). Have any steps to reproduce?

**Comment by neVERberleRfellerER** [comment]
@sebmck It is caused by bluebirdCoroutines transformer. Name of the async function is removed.

**Comment by sebmck** [comment]
Relevant then #552.

**Comment by sebmck** [comment]
Although that's only for `FunctionDeclaration`s, might be able to get away with retaining the name for `FunctionExpression`s. Not entirely sure how `bluebird.coroutine` works under the hood but I assume it should work fine `yield`ing from the generator function.

**Comment by neVERberleRfellerER** [comment]
@sebmck Yes. However, preserving name is not a solution either - bad things would happen. Quick solution that comes to my mind:
```
var arr = []; // something
arr.map(async function f() {
  return await f();
});
```
can be transformed to something like this
```
var arr = []; // something
var _coroutine1;
arr.map(coroutine1 = _bluebird.coroutine(function* () {
  return yield f();
}));
```

**Comment by benjamingr** [comment]
cc @petkaantonov 

ps. Using async functions without bluebird coroutines is kind of a nightmare :D

**Comment by sebmck** [comment]
@neVERberleRfellerER I'm confused by what you're trying to demonstrate with that example.

**Comment by petkaantonov** [comment]
If you name a generator function, the name will reference the raw generator function not the one returned by `.coroutine()`

**Comment by neVERberleRfellerER** [comment]
@sebmck Sorry, I have messed names of everything. There is better example:
```javascript
var arr = [Promise.resolve(1), Promise.resolve(2)];
var res = bluebird.map(arr, async function f(x) {
  if (x % 2 === 0) {
    return await f(x + 1);
  } else {
    return x;
  }
});
res.then(function (elems) { console.log(elems); });
```
work if transformed into
```javascript
var arr = [Promise.resolve(1), Promise.resolve(2)];
var f; // this is added
var res = bluebird.map(arr, f = _bluebird.coroutine(function* (x) { // notice "f = "
  if (x % 2 === 0) {
    return yield f(x + 1); // now f is defined and coroutine itself
  } else {
    return x;
  }
}));
res.then(function (elems) {
  console.log(elems);
});
```

**Comment by sebmck** [comment]
@petkaantonov Yep. My question was whether you can `yield` to the raw generator or if it had to be to the returned `coroutine`.

**Comment by sebmck** [comment]
@neVERberleRfellerER That wouldn't work because then `f` leaks into the parent scope when the `f` binding should be local to the function.

**Comment by petkaantonov** [comment]
@sebmck You can `coroutine.addYieldHandler()` for raw generators that wraps them in coroutines but that requires additional runtime code and is much slower than referring to the existing coroutine.

**Comment by neVERberleRfellerER** [comment]
@sebmck This is the reason, why I have renamed it to generated name (_coroutine1) in original example.
Also, you can not simply yield generator itself. he delegation could work, probably.

**Comment by sebmck** [comment]
@petkaantonov Yeah, I'll just have to rewrite all the references then. Thanks for the clarification!

@neVERberleRfellerER Ah right, from your examples it wasn't obvious that it was a UID.

**Comment by benjamingr** [comment]
Thanks for the quick fix and sorry for the less-than-stellar bug report, I ran into it during a project that didn't compile so I didn't really have time to elaborate correctly (figured I would later and that a partial report is better than nothing).

**Comment by sebmck** [comment]
All good. Fixed as of 4.7.4, thanks!
